server/include/bge/ComponentManager.h

```diff
@@ -0,0 +1,49 @@
+#pragma once
+#include <map>
+#include <array>
+#include <vector>
+#include "Entity.h"
+
+namespace bge {
+
+    template <typename ComponentType>
+    class ComponentManager {
+        public:
+            ComponentManager() {
+                componentDataStorage.reserve(256);
+            }
+            int add(Entity e, ComponentType c) {
+                int newIndex = componentDataStorage.size();
+                componentDataStorage.push_back(c);
+
+                entityMap[e.id] = newIndex;
+                componentMap[newIndex] = e.id;
+
+                return newIndex;
+            }
+            void remove(Entity e) {
+                int currSize = componentDataStorage.size();
+                int delIndex = entityMap[e.id];
+                componentDataStorage[delIndex] = componentDataStorage[currSize];
```

Comment:
If currSize is the length of the vector, won't this be out of bounds? I feel like currSize-1 makes more sense here.

Reply:
Yeah, you're correct. I've pushed a fix for this, setting currSize = componentDataStorage.size()-1;

Reply:
Sounds good